docs/integrations/channels.md

```diff
@@ -0,0 +1,209 @@
+
+A more complex use case is when you want to integrate several ASGI applications
+on different URLs and protocols, like what is described in the
+[Channels documentation](https://channels.readthedocs.io/en/stable/topics/protocols.html).
+
+An example of this would be:
+
+```python
+import os
+
+from django.core.asgi import get_asgi_application
+from django.urls import re_path
+from strawberry.channels import GraphQLWSConsumer, GraphQLHTTPConsumer
+from channels.routing import ProtocolTypeRouter, URLRouter
+from channels.auth import AuthMiddlewareStack
+
+os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'mysite.settings')
+django_asgi_app = get_asgi_application()
+
+# Import your Strawberry schema after creating the django ASGI application
+# This ensures django.setup() has been called before any ORM models are imported
+# for the schema.
+from mysite.graphql import schema
+from third_party import third_party_asgi_app
+from multichat import ChatApp
+
+application = ProtocolTypeRouter({
+    "http": URLRouter([
+        re_path("^graphql", AuthMiddlewareStack(GraphQLHTTPConsumer.as_asgi(schema=schema))),
+        re_path("^api", django_asgi_app),
+        # This might be eg a static file server or something
+        re_path("^", third_party_asgi_app),
+    ]),
+    "websocket": URLRouter([
+        re_path("^graphql", AuthMiddlewareStack(GraphQLWSConsumer.as_asgi(schema=schema))),
+        re_path("^chat", ChatApp.as_asgi()),
+    ]),
+})
+```
+
+This example demonstrates some ways that Channels can be set up to handle
+routing. A very common scenario will be that you want user and session
+information inside the GraphQL context, which the AuthMiddlewareStack wrapper
+above will provide. It might be apparent by now, there's no reason at all why
+you couldn't run a Channels server without any Django ASGI application at all.
+However, take care to ensure you run `django.setup()` instead of
+`get_asgi_application()`, if you need any Django ORM or other Django features in
+Strawberry.
+
+## Channel Layers
+
+The Context for Channels integration includes the consumer, which has an
+instance of the channel layer and the consumer's channel name. Here's an example
+of how this can be used in the schema to provide subscriptions to events
+generated by background tasks or the web server, even if these are executed in
+other threads, processes, or even other servers if you are using a Layer backend
+like Redis or RabbitMQ.
+
+To set this up, you'll need to make sure Channel Layers is configured as per the
+[documentation](https://channels.readthedocs.io/en/stable/topics/channel_layers.html).
+
+Then you'll want to add a subscription that accesses the channel layer and joins
+one or more broadcast groups.
+
+Since listening for events and passing them along to the client is a common use case,
+the base consumer provides a high level API for that using a generator pattern.
+
+For example, a basic chat room implementation might look like this:
+
```

Comment:
```suggestion
```py
from typing import AsyncGenerator, List

from channels.layers import get_channel_layer
import strawberry
from strawberry.types import Info


@strawberry.input
class ChatRoom:
    room_name: str


@strawberry.type
class ChatRoomMessage:
    room_name: str
    current_user: str
    message: str


@strawberry.type
class Mutation:
    @strawberry.mutation
    async def send_chat_message(
        self,
        room: ChatRoom,
        message: str,
    ) -> None:
        channel_layer = get_channel_layer()
        await channel_layer.group_send(
            f"chat_{room.room_name}",
            {
                "type": "chat.message",
                "room_id": room.room_name,
                "message": message,
            },
        )


@strawberry.type
class Subscription:
    @strawberry.subscription
    async def join_chat_rooms(
        self,
        info: Info,
        rooms: List[ChatRoom],
        user: str,
    ) -> AsyncGenerator[ChatRoomMessage, None]:
        """Join and subscribe to message sent to the given rooms."""
        ws = info.context.ws
        channel_layer = ws.channel_layer
        for room in (rooms := ["chat_%s" % room.room_name for room in rooms]):
            # Join room group
            await channel_layer.group_add(room, ws.channel_name)

        for room in rooms:
            await channel_layer.group_send(
                room,
                {
                    "type": "chat.message",
                    "room_id": room,
                    "message": f"Hey there my name is {user}!",
                },
            )

        async for message in ws.channel_listen("chat.message", groups=rooms):
            yield ChatRoomMessage(
                room_name=message["room_id"], message=message["message"], current_user=user
            )
```
```